src/sidecar/connections.ts

```diff
@@ -367,3 +316,151 @@ async function discoverSchemaRegistry(): Promise<string | undefined> {
   logger.debug("Discovered Schema Registry REST proxy port", { schemaRegistryPort });
   return `http://localhost:${restProxyPort}`;
 }
+
+/**
+ * Definition for predicate functions passed into ConnectionStateWatcher.waitForConnectionUpdate() to determine
+ * if a connection update should be considered for further processing. Allows for filtering out
+ * updates that are not relevant to the caller w/o gaps where there could be 0 observers and therefore
+ * missed updates.
+ * */
+export type ConnectionUpdateFilter = (connection: Connection) => boolean;
+
+export class ConnectionStateWatcher {
+  // Singleton instance
+  private static instance: ConnectionStateWatcher;
+
+  // connection id -> pair of (most recent Connection announcement, event emitter to notify observers)
+  private connectionStates: Map<ConnectionId, SingleConnectionState> = new Map();
+
+  public static getInstance(): ConnectionStateWatcher {
+    if (!ConnectionStateWatcher.instance) {
+      ConnectionStateWatcher.instance = new ConnectionStateWatcher();
+    }
+    return ConnectionStateWatcher.instance;
+  }
+
+  private constructor() {
+    // Register handleConnectionUpdateEvent as a listener for CONNECTION_EVENT websocket messages
+    WebsocketManager.getInstance().subscribe(
+      MessageType.CONNECTION_EVENT,
+      this.handleConnectionUpdateEvent.bind(this),
+    );
+  }
+
+  /**
+   * Forget any prior state for the given connection ID by resetting any map
+   * entry's mostRecentState to null. Needed when, say, a direct connection has
+   * been updated and we want to guarantee to wait for the next update to be
+   * sure it's stable.
+   */
+  purgeCachedConnectionState(connectionId: ConnectionId): void {
+    const singleConnectionState = this.connectionStates.get(connectionId);
+    if (singleConnectionState) {
+      singleConnectionState.mostRecentState = null;
+    }
+  }
+
+  /**
+   * Get the most recent Connection description for the given connection ID.
+   * Will return null if no CONNECTION_EVENT websocket messages have been
+   * received for this connection.
+   */
+  getLatestConnectionUpdate(connectionId: ConnectionId): Connection | null {
```

Comment:
May be used when we convert the ccloud auth status poller over to instead consult ConnectionStateWatcher. Otherwise if not then we should get rid of this -- not used in this branch in the final spelling.